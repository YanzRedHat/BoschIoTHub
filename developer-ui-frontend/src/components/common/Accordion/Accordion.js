/*
 * Copyright 2018 Bosch Software Innovations GmbH ("Bosch SI"). All rights reserved.
 */
import React, { Component, Fragment } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import Measure from "react-measure";
import throttle from "lodash.throttle";

export const FOOTER_HEIGHT = 40;

/* eslint-disable react/no-multi-comp */
// AccordionContainer constrains the dimensions if the maximum height gets exceeded
const AccordionContainer = styled.div`
  display: flex;
  max-height: 100%;
  width: 100%;
  flex-direction: column;
  list-style-type: none;
  margin: 0;
  padding: 0;
`;

// AccordionInner is the real dimension of the Accordion
const AccordionInner = styled.span`
  box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.12), 0 1px 2px 0 rgba(0, 0, 0, 0.24);
  background-color: #ccc;
  /* Prevent blurryness */
  -webkit-font-smoothing: subpixel-antialiased;
  backface-visibility: hidden;
  transform: translateZ(0);
  ${props => props.snapFooter && `margin-bottom: ${FOOTER_HEIGHT}px`};
`;

const FixedFooterPortal = styled.div`
  position: fixed;
  pointer-events: ${props => (props.snapFooter ? "auto" : "none")};
  margin-top: ${props => props.containerHeight - FOOTER_HEIGHT}px;
  height: 4.2rem;
  width: ${props => props.containerWidth}px;
`;

// Create a context to expose state that is global to the whole componenent (asField, name, leadingCheckbox)
// to all children
export const AccordionContext = React.createContext();
// Also create a HOC as simple API to wrap the children
export const withAccordionContext = WrappedComponent => props => (
  <AccordionContext.Consumer>
    {({ snapFooter, containerHeight }) => (
      <WrappedComponent
        {...props}
        containerHeight={containerHeight}
        snapFooter={snapFooter}
      />
    )}
  </AccordionContext.Consumer>
);

class Accordion extends Component {
  constructor(props) {
    super(props);
    this.state = {
      // Measure the visible dimensions of the container
      outerDimensions: {
        height: -1,
        width: -1
      },
      // ...and the height used by the accordion
      innerDimensions: {
        height: -1
      },
      snapFooter: false
    };
    // Debounce expensive calculations and prevent jumping snapFooter in states near the threshold
    this.calculateRemainingSpace = throttle(this.calculateRemainingSpace, 10, {
      trailing: true
    }).bind(this);
  }

  calculateRemainingSpace() {
    const containerHeight = this.state.outerDimensions.height;
    const usedHeight = this.state.innerDimensions.height;
    const diffHeight = containerHeight - usedHeight;
    let toggleFooter;
    toggleFooter === this.state.snapFooter
      ? diffHeight >= FOOTER_HEIGHT
      : diffHeight < 0;
    toggleFooter && this.setState(state => ({ snapFooter: !state.snapFooter }));
  }

  render() {
    const { children } = this.props;
    const { snapFooter } = this.state;
    const containerHeight = this.state.outerDimensions.height;
    const containerWidth = this.state.outerDimensions.width;
    return (
      <Fragment>
        <AccordionContext.Provider value={{ snapFooter, containerHeight }}>
          <Measure
            client
            onResize={contentRect => {
              this.setState({
                outerDimensions: contentRect.client
              });
            }}
          >
            {measure => (
              <AccordionContainer innerRef={measure.measureRef}>
                <Measure
                  scroll
                  onResize={contentRect => {
                    this.setState(
                      {
                        innerDimensions: { height: contentRect.scroll.height }
                      },
                      this.calculateRemainingSpace
                    );
                  }}
                >
                  {({ measureRef }) => (
                    <AccordionInner
                      snapFooter={snapFooter}
                      innerRef={measureRef}
                    >
                      {children}
                    </AccordionInner>
                  )}
                </Measure>
              </AccordionContainer>
            )}
          </Measure>
        </AccordionContext.Provider>
        <FixedFooterPortal
          id="fixed-acc-footer-modal"
          snapFooter={snapFooter}
          containerHeight={containerHeight}
          containerWidth={containerWidth}
        />
      </Fragment>
    );
  }
}
Accordion.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node)
  ])
};

export default Accordion;
