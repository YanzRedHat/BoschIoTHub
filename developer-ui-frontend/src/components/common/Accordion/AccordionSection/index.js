/*
 * Copyright 2018 Bosch Software Innovations GmbH ("Bosch SI"). All rights reserved.
 */
export { default } from "./AccordionSection";
export { default as AccordionSectionHeader } from "./AccordionSectionHeader";
export { default as AccordionSectionBody } from "./AccordionSectionBody";
