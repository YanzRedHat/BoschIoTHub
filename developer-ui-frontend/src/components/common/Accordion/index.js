/*
 * Copyright 2018 Bosch Software Innovations GmbH ("Bosch SI"). All rights reserved.
 */
export { default } from "./Accordion";
export { default as AccordionFixedFooter } from "./AccordionFixedFooter";
