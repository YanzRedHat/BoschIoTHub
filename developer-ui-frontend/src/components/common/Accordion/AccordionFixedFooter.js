/*
 * Copyright 2018 Bosch Software Innovations GmbH ("Bosch SI"). All rights reserved.
 */
import React from "react";
import ReactDOM from "react-dom";
import styled from "styled-components";
import { withAccordionContext } from "./Accordion";

const AccordionFixedFooterStyled = styled.div`
  &:not(.adding-mode) {
    cursor: pointer;
  }
  position: relative;
  z-index: 1;
  display: flex;
  height: 4rem;
  align-content: center;
  background: #fff;
  border: 1px solid #ccc;
  font-weight: 500;
  padding: 0 1rem;
  box-sizing: border-box;
  color: #757575;
  align-items: center;
  justify-content: flex-start;

  /*
  This could be used to indicate that one can scroll further
  &:before {
    content: "";
    position: absolute;
    top: -100%;
    width: 100%;
    height: 100%;
    background: radial-gradient(
      ellipse at bottom,
      rgba(0, 0, 0, 1) 0%,
      rgba(0, 0, 0, 0) 75%,
      rgba(0, 0, 0, 0) 100%
    );
  } */
`;
/* eslint-disable react/prop-types */
const AccordionFixedFooter = ({ snapFooter, ...props }) =>
  snapFooter ? (
    ReactDOM.createPortal(
      <AccordionFixedFooterStyled snapFooter={snapFooter} {...props} />,
      document.getElementById("fixed-acc-footer-modal")
    )
  ) : (
    <AccordionFixedFooterStyled snapFooter={snapFooter} {...props} />
  );

export default withAccordionContext(AccordionFixedFooter);
