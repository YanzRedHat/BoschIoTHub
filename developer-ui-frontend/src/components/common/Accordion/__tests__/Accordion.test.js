/*
 * Copyright 2018 Bosch Software Innovations GmbH ("Bosch SI"). All rights reserved.
 */
import React from "react";
import { shallow } from "enzyme";
import Accordion from "components/common/Accordion/Accordion.js";

describe("Accordion", () => {
  it("should be defined", () => {
    expect(Accordion).toBeDefined();
  });
  it("should render correctly", () => {
    const tree = shallow(<Accordion />);
    expect(tree).toMatchSnapshot();
  });
});
