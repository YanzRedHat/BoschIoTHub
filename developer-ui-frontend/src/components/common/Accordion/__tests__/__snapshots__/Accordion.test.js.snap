// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion should render correctly 1`] = `
<Fragment>
  <ContextProvider
    value={
      Object {
        "containerHeight": -1,
        "snapFooter": false,
      }
    }
  >
    <_class
      client={true}
      onResize={[Function]}
    >
      <Component />
    </_class>
  </ContextProvider>
  <styled.div
    containerHeight={-1}
    containerWidth={-1}
    id="fixed-acc-footer-modal"
    snapFooter={false}
  />
</Fragment>
`;
